File: Add-Two-Numbers/ATN.py
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


def checkTen(n, f):
    flag = f
    temp = n
    while (n is not None):
        n.val += flag
        if n.val >= 10:
            n.val -= 10
            flag = 1
        else:
            flag = 0
        temp = n
        n = n.next

    if flag == 1:
        temp.next = ListNode(1)


class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """

        flag = 0
        result = ListNode(flag)
        temp = result
        last = None

        while(l1 is not None and l2 is not None):
            temp.val += l1.val + l2.val
            if temp.val >= 10:
                temp.val -= 10
                flag = 1
            else:
                flag = 0
            temp.next = ListNode(flag)
            last = temp
            temp = temp.next
            l1 = l1.next
            l2 = l2.next

        if l1 is None:
            if l2 is not None:
                last.next = l2
                checkTen(l2, flag)
            else:
                if flag == 1:
                    last.next = ListNode(flag)
                else:
                    last.next = None
            return result

        if l2 is None:
            last.next = l1
            checkTen(l1, flag)

        return result


'''
test:
(1)
(9,2)
-> (0,3)

(2)
(8,9,9)
->(0,0,0,1)

(5)
(5)
->(0,1)

'''

File: Add-Two-Numbers/ATN.rs
#[derive(Debug)]
struct ListNode {
    val: i32,
    next: Option<Box<ListNode>>,
}

impl Clone for ListNode {
    fn clone(&self) -> ListNode {
        let mut result = ListNode::new(self.val);
        result.next = self.next.clone();
        return result;
    }
}

impl ListNode {
    fn new(v: i32) -> Self {
        ListNode { val: v, next: None }
    }

    fn check_ten(&mut self, flag: i32) {
        let mut inner_flag = flag;
        self.val += flag;
        if self.val >= 10 {
            self.val -= 10;
            inner_flag = 1;
        } else {
            inner_flag = 0;
        }

        match self.next {
            Some(ref mut n) => n.check_ten(inner_flag),
            None => {
                if inner_flag == 1 {
                    self.next = Some(Box::new(ListNode::new(inner_flag)));
                }
            }
        }
    }
}

fn add_two_numbers(a: ListNode, b: ListNode, flag: i32) -> ListNode {
    let mut inner_flag = flag;
    let mut result = ListNode::new(a.val + b.val);

    if a.next.is_some() && b.next.is_some() {
        result.val += inner_flag;
        if result.val >= 10 {
            result.val -= 10;
            inner_flag = 1;
        } else {
            inner_flag = 0;
        }

        let inner_a = Box::leak(a.next.unwrap()).clone();
        let inner_b = Box::leak(b.next.unwrap()).clone();
        result.next = Some(Box::new(add_two_numbers(inner_a, inner_b, inner_flag)))
    } else if a.next.is_none() && b.next.is_some() {
        let temp = Box::leak(b.next.unwrap()).clone();
        result.next = Some(Box::new(temp));
        result.check_ten(inner_flag)
    } else if b.next.is_none() && a.next.is_some() {
        let temp = Box::leak(a.next.unwrap()).clone();
        result.next = Some(Box::new(temp));
        result.check_ten(inner_flag)
    } else if a.next.is_none() && b.next.is_none() {
        result.check_ten(inner_flag);
    }

    return result;
}

fn main() {
    let a = ListNode::new(1);
    let mut b = ListNode::new(9);
    b.next = Some(Box::new(ListNode::new(2)));

    println!("{:?}", add_two_numbers(a, b, 0));

    let c = ListNode::new(2);
    let mut d = ListNode::new(8);
    let mut d1 = ListNode::new(9);
    let mut d2 = ListNode::new(9);
    d1.next = Some(Box::new(d2));
    d.next = Some(Box::new(d1));

    println!("{:?}", add_two_numbers(c, d, 0));
}
